login: warn about --password on command line

tych0 opened the pull request:
This isn't safe, since every commands arguments are available via
/proc/<pid>/cmdline.

Let's print a nasty warning and then add a --password-file option, so
people can use a password file instead if they want automated access.

Signed-off-by: Tycho Andersen <tycho@tycho.ws>


![cute dog](http://justcuteanimals.com/wp-content/uploads/2016/07/cute-funny-pug-life-shades-summer-dogs-puppy-animal-pictures.jpg)

cpuguy83 commented:
Needs a rebase.

tych0 commented:
Done, looks like the previous merge broke the build, so I fixed that too, let me know if my fix is wrong; I don't know anything about this code :)

tych0 commented:
Hmm. And it seems the linter is complaining about this... but if we try it the way it was (the way the linter is suggesting), it doesn't compile.

dnephin commented on cli/command/registry/login.go:
```diff
@@ -47,6 +51,34 @@ func runLogin(dockerCli command.Cli, opts loginOptions) error {
 	clnt := dockerCli.Client()
 
+	if opts.password != "" {
+		fmt.Fprintf(os.Stderr, "Using --password via the CLI is insecure. Please use --password-file.\n")
```
`os.Stderr` should be `dockerCli.Err()`

dnephin commented:
> but if we try it the way it was (the way the linter is suggesting), it doesn't compile.

It does compile, CI was green. Are you building outside of a container with an older version of GO maybe?

dnephin commented on cli/command/registry/login.go:
```diff
@@ -47,6 +51,34 @@ func runLogin(dockerCli command.Cli, opts loginOptions) error {
+		fmt.Fprintf(os.Stderr, "Using --password via the CLI is insecure. Please use --password-file.\n")
+	}
+
+	if opts.passwordFile != "" {
```
We should also support reading from stdin if `passwordFile == "-"`. There is code to handle that from an arg in other commands. I'm not sure if we have a function for it yet, or if maybe we could extract one.

dnephin commented on cli/command/registry/login.go:
```diff
@@ -47,6 +51,34 @@ func runLogin(dockerCli command.Cli, opts loginOptions) error {
+		 * conforms to the principle of least surprise, but I could be
+		 * wrong :)
+		 */
+		if contents[len(contents)-1] == '\n' {
```
Anytime I have a long comment like this I think it's a good idea to split it into a new function. That way the comment can serve as a docstring on the function instead of cluttering up the logic within a function. Even if the function is only 3 lines, that's fine.

Maybe something like `normalizePasswordFileContent()` ?

dnephin commented:
I'm looking into the other 2 linter errors. I don't see them failing on master.

dnephin commented:
Linter errors are fixed in #221 if you want to rebase on that commit (or just wait until it gets merged). There's some explanation of the problem as well.

tych0 commented:
On Tue, Jun 20, 2017 at 9:40 PM, Daniel Nephin <notifications@github.com>
wrote:

> but if we try it the way it was (the way the linter is suggesting), it
> doesn't compile.
>
> It does compile, CI was green. Are you building outside of a container
> with an older version of GO maybe?
>

I'm just typing "make" :). My host go version is 1.7.4.


> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/cli/pull/218#issuecomment-309954205>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAv612gVBfkO3vY_T-GdtvVU4OzGvozOks5sGJDCgaJpZM4OAS0D>
> .
>

dnephin commented:
You can use `make -f docker.Makefile ...`  to run in a conatiner: https://github.com/docker/cli#development

tych0 commented:
Thanks; seems that still didn't work for me (it didn't use the current version, where there was an obvious syntax error, but instead compiled successfully). Anyway, I dropped that commit and addressed all your feed back. I'm happy to rebase once #221 goes in, just ping me.

dnephin commented on cli/command/registry/login.go:
```diff
@@ -39,14 +41,62 @@ func NewLoginCommand(dockerCli command.Cli) *cobra.Command {
+ * For users that do have a \n as the last character of their password, they
+ * need to store it as \n\n. I think this conforms to the principle of least
+ * surprise, but I could be wrong :)
+ */
```
I believe the go convention is to use `//` single line comments for everything instead of `/* ... */`

tych0 commented:
Fixed, thanks!

tych0 commented:
I've just rebased this now that #221 is merged, so this should be good to go as well.

n4ss commented:
@vdemeester big improvement on the discussion we had. I'll open an issue to make sure that users simply don't provide passwords on the cli

tych0 commented:
The CLI library doesn't seem to like -pf:

```
panic: shorthand is more than one character

goroutine 1 [running]:
panic(0xd24a00, 0xc4203b1050)
	/usr/lib/go-1.7/src/runtime/panic.go:500 +0x1a1
github.com/docker/cli/vendor/github.com/spf13/pflag.(*FlagSet).AddFlag(0xc420400900, 0xc420410000)
	/home/tycho/packages/go/src/github.com/docker/cli/vendor/github.com/spf13/pflag/flag.go:738 +0x61f
github.com/docker/cli/vendor/github.com/spf13/pflag.(*FlagSet).VarPF(0xc420400900, 0x135e120, 0xc42035d1f0, 0xe85c29, 0xd, 0xe7974b, 0x2, 0xeb2e26, 0x34, 0xc420403f40)
	/home/tycho/packages/go/src/github.com/docker/cli/vendor/github.com/spf13/pflag/flag.go:706 +0x141
github.com/docker/cli/vendor/github.com/spf13/pflag.(*FlagSet).VarP(0xc420400900, 0x135e120, 0xc42035d1f0, 0xe85c29, 0xd, 0xe7974b, 0x2, 0xeb2e26, 0x34)
	/home/tycho/packages/go/src/github.com/docker/cli/vendor/github.com/spf13/pflag/flag.go:712 +0x8e
github.com/docker/cli/vendor/github.com/spf13/pflag.(*FlagSet).StringVarP(0xc420400900, 0xc42035d1f0, 0xe85c29, 0xd, 0xe7974b, 0x2, 0x0, 0x0, 0xeb2e26, 0x34)
	/home/tycho/packages/go/src/github.com/docker/cli/vendor/github.com/spf13/pflag/string.go:42 +0xab
github.com/docker/cli/cli/command/registry.NewLoginCommand(0x13620e0, 0xc4201cf740, 0xc4203f8480)
	/home/tycho/packages/go/src/github.com/docker/cli/cli/command/registry/login.go:44 +0x2e1
github.com/docker/cli/cli/command/commands.AddCommands(0xc4203ae480, 0xc4201cf740)
	/home/tycho/packages/go/src/github.com/docker/cli/cli/command/commands/commands.go:51 +0x175
main.newDockerCommand(0xc4201cf740, 0xc42002e008)
	/home/tycho/packages/go/src/github.com/docker/cli/cmd/docker/docker.go:68 +0x4ca
main.main()
	/home/tycho/packages/go/src/github.com/docker/cli/cmd/docker/docker.go:166 +0xbd
```

so I've just dropped the short flag for now, and made them mutually exclusive.

thaJeztah commented:
Looks like I'm a bit late to the party; I'm

:+1: on a big red warning when using `--password`
:-1: on adding a `password-file` option

Having a password file means: hey! you should store your password in a plain-text file on your machine!

I don't think we should promote this

What if, instead, we allow the password to be piped in? Looks like it's currently not supported;

```bash
# or whatever means to pipe the password
$ echo $SECRET | docker login myname
Error: Cannot perform an interactive login from a non TTY device
```

tych0 commented:
Done, see #270 and #271 (note that 271 has 270's commit in it too, because I update the message there).

jikuma commented:
please dont just have --password-file options but also have environment variables like DOCKER_ServerURL, DOCKER_Username and DOCKER_Secret. This can help in atomating the docker login process.

tych0 commented:
Putting secrets in environment variables is just as unsafe as putting them in the command line (the task's initial environment is available via /proc/pid/environ). There are some other ways this leaks too: https://diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/

So the patches will only support taking passwords via stdin.
